qiskit/algorithms/phase_estimators/ipe.py

```diff
@@ -157,8 +179,14 @@ def estimate(
 
         Returns:
             Estimated phase in an IterativePhaseEstimationResult object.
-        """
 
+        Raises:
+            AlgorithmError: If neither sampler nor quantum instance is provided.
+        """
+        if self._sampler is None and self._quantum_instance is None:
+            raise AlgorithmError(
+                "Neither a sampler nor a quantum instance was provided. Please provide one of them."
+            )
```

Comment:
At present none of the algos here seem to have a way to update quantum_instance so having None on the constructor, aside from it being helpful now we are adding Sampler, I would have said was in error since there is no way to set it later. Maybe after we drop QI we can remove None from the Sampler and require it upfont. For now None on each as default allows one to be passed, and as such perhaps we should move this test to the constructor to do the check as early as possible.

Reply:
Good idea, thanks!